### PR TITLE
DA-825 : article on pareto selection

### PR DESCRIPTION
**Pull request by sandipmukherjee**
I have created an article on pareto selection in our tech blog.
Please review and suggest changes before publishing it.

### CODE REVIEW

File: _data/authors.yml [hunk 1]
```diff
@@ -61,3 +61,10 @@ oleg_hein:
     role: Senior Frontend Engineer
     avatar: authors/oleg-hein.jpg
     bio: Oleg loves building frontends that shine - visually and technically.
+sandip_mukherjee:
+    name: Sandip Mukherjee
+    email: tech@comtravo.com
+    web: https://www.github.com/sandipmukherjee
+    role: Machine Learning Engineer
+    avatar: authors/sandipmukherjee.jpeg
+    bio: Sandip works in NLP pipeline and automation system in comtrao to automate email requests.
```
Comment (sebastianmika) [inline _data/authors.yml:70]: Pls. choose a neutral description about yourself that is not directly related to your working at Comtravo. Can by a but funny as well

File: _posts/2021-07-15-pareto_selection.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+description: Pareto optimization for ranking of search results
+usemathjax: true
+author: sandip_mukherjee
+image:
+  teaser: 2019_06_19/teaser.jpg
+  feature: 2019_06_19/feature.jpg
+
+---
+
+In comtravo, We have email requests coming from the user for a travel itinerary. The NLP engine converts that email to a structured request and our automation module receives the request for further offer generation. An example of a structured hotel request is below.
```
Comment (davidsbatista) [inline _posts/2021-07-15-pareto_selection.md:18]: typo: `Comtravo`

File: _posts/2021-07-15-pareto_selection.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+```
+
+Based on such structured representation of a customer request, the automation module searches for supply, e.g. flights, trains, or hotels, in accordance with the request using third party APIs. Depending on the vertical and the request, anything between zero and a few thousands of search results can be returned from the supply side(s).
+
+To generate an offer, the automation must now rank those search results and find the best possible options. The best matching options are then send to the customer:
+
+<center>
+<figure>
+  <img style="width: 70%; height: 70%" src="/images/2021_07_15/automation_flow.png">
```
Comment (sebastianmika) [inline _posts/2021-07-15-pareto_selection.md:51]: Maybe fewer steps would do: Request -> Search -> Ranking -> Offer?

**Comment by sandipmukherjee** [comment]
@sebastianmika  Can you please have a final look. I updated according to your suggestions.

**Review by sebastianmika — APPROVED** [review]
Generally LGTM; maybe the flow diagram could be a bit better looking (and less detailed)

From my perspective good enough to go, but maybe lets get feedback from David and Scott at least. 

There is no rush to release today. Anything before August 1st is fine I guess.

File: _posts/2021-07-15-pareto_selection.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+description: Pareto optimization for ranking of search results
+usemathjax: true
+author: sandip_mukherjee
+image:
+  teaser: 2019_06_19/teaser.jpg
+  feature: 2019_06_19/feature.jpg
+
+---
+
+At Comtravo, amongst others, we process incoming requests with travel itineraries. One way or the another, such requests are converted to a structured format and an automation module receives that request to spark automatic generation of offers. Such automatic offers are later reviewed by agents or - if they pass our QA scoring - directly send to customers.
```
Comment (davidsbatista) [inline _posts/2021-07-15-pareto_selection.md:18]: `if they pass our QA scoring` -> maybe -> `if they pass our quality estimation scoring` ?

File: _posts/2021-07-15-pareto_selection.md (hunk 3 above)
Comment (davidsbatista) [reply _posts/2021-07-15-pareto_selection.md:51]: just my OCD: center everything both vertically and horizontally, bigger/fatter arrows

also agree with Sebastian's suggestion, but maybe `Pareto Ranking` instead of just Ranking...or maybe not, Pareto word comes later in the post....

Comment (davidsbatista) [reply _posts/2021-07-15-pareto_selection.md:51]: Give the figure a number maybe? `Figure: 1 Automation Flow`

File: _posts/2021-07-15-pareto_selection.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+* Any point in the Pareto front is considered *Pareto Optimal*. By moving along the edge we can minimize price at expense of duration, or minimize duration at expense of price, but we cannot minimize both.
+* In this case, both `Train 1` and `Train 2` will be Pareto optimal points. Hence they will rank higher than `Train 3`. Ranking between `Train 1` and `Train 2` - without any further conditions like in this example - is arbitrary.
+* `Train 3` is called *Pareto Inefficient* as there exists at least one more point which is better than `Train 3` in all objectives.
+* In our algorithm, we put Pareto optimal points first and then append Pareto inefficient points to the ranking. We do not exclude Pareto inefficient points - if there are very few (or no) Pareto optimal points we might still offer them.
+
+### Example 2 - One Pareto Optimal Result
+
+In this example we will minimize one objective **Price** and try to satisfy one constraint **FlexibleTicketConstraint**.
```
Comment (davidsbatista) [inline _posts/2021-07-15-pareto_selection.md:101]: maybe a parentheses or quick note, e.g. - `"FlexibleTicketConstraint", meaning the customer wants a Deutsche Bahn's flexible ticket`

File: _posts/2021-07-15-pareto_selection.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+In the end each item will either be marked Pareto optimal or Pareto inefficient. The algorithm will rank the Pareto optimal points higher than others.
+
+## Conclusion
+
+Pareto selection works very well for this use case, especially when the request has many constraints since the algorithm is designed to handle multi-objective and multi constraints optimization problem.
+
+Of course in practice we still need a way to rank amongst the Pareto optimal and Pareto inefficent points.
+This we achieve by specifying a precedence between objectives and sorting among the Pareto optimal points accordingly. How we do this in detail is, however, material for another article.
+
+Our special thanks to Gabriele Lanaro for contribution in ideation and implementation of the Pareto selection for ranking.
```
Comment (davidsbatista) [inline _posts/2021-07-15-pareto_selection.md:164]: let's maybe give Gabi some in-links ?

`[Gabriele Lanaro](http://gabrielelanaro.github.io/)`

**Comment by davidsbatista** [comment]
Looks good! Just some small suggestions! Well done Sandip :)

**Review by davidsbatista — APPROVED** [review]
